File: dining_philosophers.cpp
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>
#include <time.h>

#define NUM_FILOSOFOS 5
#define TIEMPO_PENSAR 2000000 // 2 segundos
#define TIEMPO_COMER 1000000  // 1 segundo

// Estados posibles de un filósofo
typedef enum
{
    PENSANDO,
    HAMBRIENTO,
    COMIENDO
} estado_filosofos;

// Variables globales
pthread_mutex_t tenedores[NUM_FILOSOFOS];
pthread_mutex_t mutex_estado;
sem_t limite_comensales; // Límite de comensales comiendo
estado_filosofos estado[NUM_FILOSOFOS];
int comensales_comiendo = 0;

// Índice del tenedor izquierdo
int izquierdo(int id_filos)
{
    return id_filos;
}

// Índice del tenedor derecho
int derecho(int id_filos)
{
    return (id_filos + 1) % NUM_FILOSOFOS;
}

void mostrar_estado(int id, const char *accion)
{
    pthread_mutex_lock(&mutex_estado);
    printf("Filósofo %d está %s | Estados: [", id, accion);
    for (int i = 0; i < NUM_FILOSOFOS; i++)
    {
        switch (estado[i])
        {
        case PENSANDO:
            printf("P");
            break;
        case HAMBRIENTO:
            printf("H");
            break;
        case COMIENDO:
            printf("C");
            break;
        }
        if (i < NUM_FILOSOFOS - 1)
            printf(" ");
    }
    printf("] Comiendo: %d\n", comensales_comiendo);
    pthread_mutex_unlock(&mutex_estado);
}

// Solución 1: Orden asimétrico para evitar interbloqueo
void tomar_tenedores_asimetrico(int id)
{
    int izq = izquierdo(id);
    int der = derecho(id);

    pthread_mutex_lock(&mutex_estado);
    estado[id] = HAMBRIENTO;
    pthread_mutex_unlock(&mutex_estado);

    mostrar_estado(id, "hambriento");

    if (id % 2 == 0)
    {
        pthread_mutex_lock(&tenedores[izq]);
        pthread_mutex_lock(&tenedores[der]);
    }
    else
    {
        pthread_mutex_lock(&tenedores[der]);
        pthread_mutex_lock(&tenedores[izq]);
    }

    pthread_mutex_lock(&mutex_estado);
    estado[id] = COMIENDO;
    comensales_comiendo++;
    pthread_mutex_unlock(&mutex_estado);

    mostrar_estado(id, "comiendo");
}

void soltar_tenedores_asimetrico(int id)
{
    int izq = izquierdo(id);
    int der = derecho(id);

    pthread_mutex_lock(&mutex_estado);
    estado[id] = PENSANDO;
    comensales_comiendo--;
    pthread_mutex_unlock(&mutex_estado);

    pthread_mutex_unlock(&tenedores[izq]);
    pthread_mutex_unlock(&tenedores[der]);

    mostrar_estado(id, "pensando");
}

// Solución 2: Semáforo para limitar comensales
void tomar_tenedores_semaforo(int id)
{
    pthread_mutex_lock(&mutex_estado);
    estado[id] = HAMBRIENTO;
    pthread_mutex_unlock(&mutex_estado);

    mostrar_estado(id, "hambriento");

    sem_wait(&limite_comensales);

    int izq = izquierdo(id);
    int der = derecho(id);

    if (izq < der)
    {
        pthread_mutex_lock(&tenedores[izq]);
        pthread_mutex_lock(&tenedores[der]);
    }
    else
    {
        pthread_mutex_lock(&tenedores[der]);
        pthread_mutex_lock(&tenedores[izq]);
    }

    pthread_mutex_lock(&mutex_estado);
    estado[id] = COMIENDO;
    comensales_comiendo++;
    pthread_mutex_unlock(&mutex_estado);

    mostrar_estado(id, "comiendo");
}

void soltar_tenedores_semaforo(int id)
{
    int izq = izquierdo(id);
    int der = derecho(id);

    pthread_mutex_lock(&mutex_estado);
    estado[id] = PENSANDO;
    comensales_comiendo--;
    pthread_mutex_unlock(&mutex_estado);

    pthread_mutex_unlock(&tenedores[izq]);
    pthread_mutex_unlock(&tenedores[der]);

    sem_post(&limite_comensales);
    mostrar_estado(id, "pensando");
}

void *filosofo_asimetrico(void *arg)
{
    int id = *((int *)arg);
    for (int i = 0; i < 3; i++)
    {
        mostrar_estado(id, "pensando");
        usleep(rand() % TIEMPO_PENSAR);
        tomar_tenedores_asimetrico(id);
        usleep(rand() % TIEMPO_COMER);
        soltar_tenedores_asimetrico(id);
    }
    printf("Filósofo %d terminó de comer\n", id);
    return NULL;
}

void *filosofo_semaforo(void *arg)
{
    int id = *((int *)arg);
    for (int i = 0; i < 3; i++)
    {
        mostrar_estado(id, "pensando");
        usleep(rand() % TIEMPO_PENSAR);
        tomar_tenedores_semaforo(id);
        usleep(rand() % TIEMPO_COMER);
        soltar_tenedores_semaforo(id);
    }
    printf("Filósofo %d terminó de comer\n", id);
    return NULL;
}

void ejecutar_asimetrico()
{
    printf("\n=== Solución con Orden Asimétrico ===\n");
    pthread_t hilos[NUM_FILOSOFOS];
    int ids[NUM_FILOSOFOS];

    for (int i = 0; i < NUM_FILOSOFOS; i++)
    {
        estado[i] = PENSANDO;
        ids[i] = i;
        pthread_create(&hilos[i], NULL, filosofo_asimetrico, &ids[i]);
    }
    for (int i = 0; i < NUM_FILOSOFOS; i++)
    {
        pthread_join(hilos[i], NULL);
    }
    printf("Solución asimétrica finalizada\n");
}

void ejecutar_semaforo()
{
    printf("\n=== Solución con Semáforo ===\n");
    pthread_t hilos[NUM_FILOSOFOS];
    int ids[NUM_FILOSOFOS];

    for (int i = 0; i < NUM_FILOSOFOS; i++)
    {
        estado[i] = PENSANDO;
        ids[i] = i;
    }
    comensales_comiendo = 0;

    for (int i = 0; i < NUM_FILOSOFOS; i++)
    {
        pthread_create(&hilos[i], NULL, filosofo_semaforo, &ids[i]);
    }
    for (int i = 0; i < NUM_FILOSOFOS; i++)
    {
        pthread_join(hilos[i], NULL);
    }
    printf("Solución con semáforo finalizada\n");
}

int main()
{
    srand(time(NULL));

    for (int i = 0; i < NUM_FILOSOFOS; i++)
    {
        pthread_mutex_init(&tenedores[i], NULL);
    }
    pthread_mutex_init(&mutex_estado, NULL);
    sem_init(&limite_comensales, 0, NUM_FILOSOFOS - 1);

    printf("Problema de los Filósofos Comensales\n");
    printf("Filósofos: %d | Cada uno comerá 3 veces\n", NUM_FILOSOFOS);

    ejecutar_asimetrico();
    sleep(2);
    ejecutar_semaforo();

    for (int i = 0; i < NUM_FILOSOFOS; i++)
    {
        pthread_mutex_destroy(&tenedores[i]);
    }
    pthread_mutex_destroy(&mutex_estado);
    sem_destroy(&limite_comensales);

    return 0;
}

File: dining_philosophers.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

const (
	NUM_FILOSOFOS = 5
	TIEMPO_PENSAR = 2000
	TIEMPO_COMER  = 1000
)

const (
	PENSANDO = iota
	HAMBRIENTO
	COMIENDO
)

var (
	tenedores          [NUM_FILOSOFOS]sync.Mutex
	estado             [NUM_FILOSOFOS]int
	mutex_estado       sync.Mutex
	comensales_comiendo = 0
	limite_comensales  = make(chan struct{}, NUM_FILOSOFOS-1)
)

// Muestra los estados de los filósofos
func mostrar_estado(id int, accion string) {
	mutex_estado.Lock()
	defer mutex_estado.Unlock()

	fmt.Printf("Filósofo %d está %s | Estados: [", id, accion)
	for i := 0; i < NUM_FILOSOFOS; i++ {
		switch estado[i] {
		case PENSANDO:
			fmt.Print("P")
		case HAMBRIENTO:
			fmt.Print("H")
		case COMIENDO:
			fmt.Print("C")
		}
		if i < NUM_FILOSOFOS-1 {
			fmt.Print(" ")
		}
	}
	fmt.Printf("] Comiendo: %d\n", comensales_comiendo)
}

// Índice del tenedor izquierdo
func izquierdo(id int) int {
	return id
}

// Índice del tenedor derecho
func derecho(id int) int {
	return (id + 1) % NUM_FILOSOFOS
}

// Solución 1: Enfoque asimétrico
func tomar_tenedores_asimetrico(id int) {
	izq := izquierdo(id)
	der := derecho(id)

	mutex_estado.Lock()
	estado[id] = HAMBRIENTO
	mutex_estado.Unlock()

	mostrar_estado(id, "hambriento")

	// Los filósofos pares toman primero el izquierdo, los impares el derecho
	if id%2 == 0 {
		tenedores[izq].Lock()
		tenedores[der].Lock()
	} else {
		tenedores[der].Lock()
		tenedores[izq].Lock()
	}

	mutex_estado.Lock()
	estado[id] = COMIENDO
	comensales_comiendo++
	mutex_estado.Unlock()

	mostrar_estado(id, "comiendo")
}

func soltar_tenedores_asimetrico(id int) {
	izq := izquierdo(id)
	der := derecho(id)

	mutex_estado.Lock()
	estado[id] = PENSANDO
	comensales_comiendo--
	mutex_estado.Unlock()

	tenedores[izq].Unlock()
	tenedores[der].Unlock()

	mostrar_estado(id, "pensando")
}

// Solución 2: Enfoque con semáforo
func tomar_tenedores_semaforo(id int) {
	mutex_estado.Lock()
	estado[id] = HAMBRIENTO
	mutex_estado.Unlock()

	mostrar_estado(id, "hambriento")

	// Espera por semáforo - limita los comensales concurrentes
	limite_comensales <- struct{}{}

	izq := izquierdo(id)
	der := derecho(id)

	// Siempre toma los tenedores en un orden consistente para evitar interbloqueos
	if izq < der {
		tenedores[izq].Lock()
		tenedores[der].Lock()
	} else {
		tenedores[der].Lock()
		tenedores[izq].Lock()
	}

	mutex_estado.Lock()
	estado[id] = COMIENDO
	comensales_comiendo++
	mutex_estado.Unlock()

	mostrar_estado(id, "comiendo")
}

func soltar_tenedores_semaforo(id int) {
	izq := izquierdo(id)
	der := derecho(id)

	mutex_estado.Lock()
	estado[id] = PENSANDO
	comensales_comiendo--
	mutex_estado.Unlock()

	tenedores[izq].Unlock()
	tenedores[der].Unlock()

	<-limite_comensales // Libera el semáforo

	mostrar_estado(id, "pensando")
}

// Filósofo con solución asimétrica
func filosofo_asimetrico(id int, wg *sync.WaitGroup) {
	defer wg.Done()

	for i := 0; i < 3; i++ {
		mostrar_estado(id, "pensando")
		time.Sleep(time.Duration(rand.Intn(TIEMPO_PENSAR)) * time.Millisecond)

		tomar_tenedores_asimetrico(id)
		time.Sleep(time.Duration(rand.Intn(TIEMPO_COMER)) * time.Millisecond)
		soltar_tenedores_asimetrico(id)
	}

	fmt.Printf("Filósofo %d terminó de comer\n", id)
}

// Filósofo con solución de semáforo
func filosofo_semaforo(id int, wg *sync.WaitGroup) {
	defer wg.Done()

	for i := 0; i < 3; i++ {
		mostrar_estado(id, "pensando")
		time.Sleep(time.Duration(rand.Intn(TIEMPO_PENSAR)) * time.Millisecond)

		tomar_tenedores_semaforo(id)
		time.Sleep(time.Duration(rand.Intn(TIEMPO_COMER)) * time.Millisecond)
		soltar_tenedores_semaforo(id)
	}

	fmt.Printf("Filósofo %d terminó de comer\n", id)
}

func ejecutar_asimetrico() {
	var wg sync.WaitGroup

	fmt.Println("\n=== Solución con Orden Asimétrico ===")

	for i := 0; i < NUM_FILOSOFOS; i++ {
		estado[i] = PENSANDO
	}
	comensales_comiendo = 0

	for i := 0; i < NUM_FILOSOFOS; i++ {
		wg.Add(1)
		go filosofo_asimetrico(i, &wg)
	}

	wg.Wait()
	fmt.Println("Solución asimétrica finalizada")
}

func ejecutar_semaforo() {
	var wg sync.WaitGroup

	fmt.Println("\n=== Solución con Semáforo ===")

	for i := 0; i < NUM_FILOSOFOS; i++ {
		estado[i] = PENSANDO
	}
	comensales_comiendo = 0

	for i := 0; i < NUM_FILOSOFOS; i++ {
		wg.Add(1)
		go filosofo_semaforo(i, &wg)
	}

	wg.Wait()
	fmt.Println("Solución con semáforo finalizada")
}

func main() {
	rand.Seed(time.Now().UnixNano())

	fmt.Println("Problema de los Filósofos Comensales")
	fmt.Printf("Filósofos: %d | Cada uno comerá 3 veces\n", NUM_FILOSOFOS)

	ejecutar_asimetrico()
	time.Sleep(2 * time.Second)
	ejecutar_semaforo()
}